### DISCUSSION TITLE
Dynamically loading pgrst.db_schemas in cloud hosted environments for non super users

### DISCUSSION THREAD

**Discussion by akshaysasidrn** [PostgREST/postgrest#2552] [Q&A] [answered]
Need some help in figuring out an approach to update postgrest server with db schema changes. 

What I'd want:
Being able to dynamically create schemas within a database and associate it with a user.
This then needs to be updated for `db_schemas` config for postgrest server to be able to handle queries scoped onto this schema with `accept-profile` and `accept-content` headers.

Issue:
The application I'm working on is self-hosted and I can't control wherein the postgres database is setup. For cloud providers such as AWS, we don't actually get a super user such that we could set parameters for db_schema. 

Trying to set parameter with:
`ALTER ROLE rolename SET pgrst.db_schemas = 'schema1,schema2'`

will result in:
```
ERROR:  permission denied to set parameter "pgrst.db_schemas"
SQL state: 42501
```

Is there any workaround for this that can be used?

**Comment by wolfgangwalther** [comment]
> Trying to set parameter with:
> `ALTER ROLE rolename SET pgrst.db_schemas = 'schema1,schema2'`
>
>will result in:
>
> ```
> ERROR:  permission denied to set parameter "pgrst.db_schemas"
> SQL state: 42501
> ```
> Is there any workaround for this that can be used?

What about `ALTER ROLE rolename IN DATABASE databasename SET pgrst.db_schemas = 'schema1,schema2'` - does that result in the same error?

**Reply by akshaysasidrn** [reply]
Yeah, got the same error.

**Comment by akshaysasidrn** [comment]
Just bouncing off ideas, not sure about how applicable it might be as I'm new to Postgres DB configuration and PostgREST.
Will it makes sense to either:
1. Have PostgREST infer db-schemas from the search path of the user? As on testing on AWS RDS Postgres, the user has permission to do this.
2. Expose an API behind some authorization on PostgREST, enabling to update db-schemas with reload.

FYI, [on AWS](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/Appendix.PostgreSQL.CommonDBATasks.Roles.html) the user created as such:
`CREATE ROLE postgres WITH LOGIN NOSUPERUSER INHERIT CREATEDB CREATEROLE NOREPLICATION VALID UNTIL 'infinity'`

**Reply by steve-chavez** [reply]
1) `db-schemas` doesn't really map to `search_path` because there might be schemas in search path that the user doesn't want to expose through the API. See [db-extra-search-path](https://postgrest.org/en/stable/configuration.html#db-extra-search-path).

2) Seems possible because we have an `/admin` endpoint that's exposed on a port different from the main API([ref](https://postgrest.org/en/stable/admin.html#health-check)). It doesn't have any auth built-in though, its security relies on only being reachable in a private network.

**Reply by akshaysasidrn** [reply]
2. It is a solution which will work for me. Adding a configurable flag which exposes such an API endpoint would help to skip the auth. And one can only enable this if the server is deployed on a private network. If such an approach sits well PostgREST, can this be made an issue to be picked up?

**Reply by steve-chavez** [reply]
>If such an approach sits well PostgREST, can this be made an issue to be picked up?

Sure, please open one.

If you'd like to give it a try, the admin server is basically this [function](https://github.com/PostgREST/postgrest/blob/b2935ef7990107642e261b25b6f30a2aed932864/src/PostgREST/Workers.hs#L293-L308)

**Reply by wolfgangwalther** [reply]
There would be no way to persist this configuration data, which is changed through the `/admin` endpoint, though. So this would probably need a much bigger framework of how to deal with configuration through `/admin` in general.

**Reply by wolfgangwalther** [reply]
What if we support something like dynamic configuration instead? I.e. instead of launching PostgREST with a config file, we load it with a command which should be executed when reloading the config file. This command can then return a dynamically created config. Having this, you could trigger a config reload via `NOTIFY` - and then have a script to create the configuration that connects to the PostgreSQL server and fetches the data whereever you store it instead of in `ALTER ROLE ... SET`. Or it takes the data from somewhere else.

The key difference: This would work across PostgREST restarts as well, because the config command can be executed on demand, while a http request can't be repeated on demand (demand by PostgREST itself).

**Reply by wolfgangwalther** [reply]
I think this could also help with jwt key rotation - didn't we have an issue somewhere discussing that?

**Reply by thomasjungblut** [reply]
also running into this right now on GCP with cloudsql, which has this feature locked-down, too. 
What's the rationale to use a server parameter here? Can't this information be retrieved from a simple config table?

**Reply by steve-chavez** [reply]
On v11.1.0,  this doesn't require superuser. See https://postgrest.org/en/stable/references/configuration.html#in-db-config

**Reply by thomasjungblut** [reply]
thanks @steve-chavez that seems to work! 

I'm only a little unclear about how to use it when changing schemas. I'm adding a new schema at runtime to PG, what are the steps required to update?

From the current documentation I would need to `create or replace function postgrest.pre_config()` entirely with the new `set_config('pgrst.db_schemas', 'schema1, schema2, schema_NEW', true)`? 

Would the configuration reload signal then call that function?

**Reply by steve-chavez** [reply, marked as answer]
@thomasjungblut Some options for the `pre-config`:

1) If your dynamic schemas names have some sort of prefix, like `tenant_`, you could do:

```sql
select 
  set_config('pgrst.db_schemas', string_agg(nspname, ','), true) 
from pg_namespace 
where nspname like 'tenant_%';

-- then reload each time you create a new schema
NOTIFY pgrst, 'reload config';
```

2) If there's no name pattern but they're created with a particular role, like `CREATE SCHEMA mine AUTHORIZATION joe`:

```sql
select 
  set_config('pgrst.db_schemas', string_agg(nspname, ','), true) 
from pg_namespace 
where nspowner = 'joe'::regrole;

-- then reload each time you create a new schema
NOTIFY pgrst, 'reload config';
```

3) Otherwise, you might need to create an additional table:

```sql
--pseudocode
create table postgrest.config (..);

select 
  set_config('pgrst.db_schemas', schemas, true) 
from postgrest.config;

-- then each time you create a new schema
--pseudocode
update postgrest.config set schemas = schemas || ', new_schema'
-- and reload
NOTIFY pgrst, 'reload config';
```

I would try 1 or 2 first to avoid extra state.

**Reply by thomasjungblut** [reply]
Amazing idea, of course we have a whole database in the back we can use for that instead. Thanks for reminding me. 
We don't have a schema prefix, but I think either the auth or the explicit table works. 

Thanks a ton for your help!
